txpipe/simulation.py

```diff
@@ -0,0 +1,544 @@
+    ]
+
+    outputs = [
+        ("photometry_catalog", HDFFile),
+        ("lens_tomography_catalog_unweighted", TomographyCatalog),
+        ("glass_cl_shells", HDFFile),
+        ("glass_cl_binned", HDFFile),
+        ("density_shells", HDFFile),
+        # TO DO: add shear maps
+    ]
+
+    config_options = {
+        "num_dens": None,
+        "zmin": 0.0,
+        "zmax": 2.0,
+        "dx": 100,
+        "bias": [1.0],
+        "shift": 1.0,
+        "contaminate": False,
+        "random_seed": 0,
+        "cl_optional_file": "none",
+        "ell_binned_min": 0.1,
+        "ell_binned_max": 5.0e5,
+        "ell_binned_nbins": 100,
+        "output_density_shell_maps":False,
+    }
+
+    def run(self):
+        """
+        Execute the TXLogNormalGlass pipeline stage
+
+        This method coordinates the execution of the GLASS-based pipeline,
+        generating simulated catalogs from lognormal fields
+
+        It calls the following sub-methods:
+        - set_z_shells
+        - generate_shell_cls #these are the C(l) used to generate the mocks
+        - generate_binned_cls #these are the C(l) the user can use for testing
+        - generate_catalogs
+        - save_catalogs
+        """
+
+        # get nside and nest info from mask
+        with self.open_input("mask", wrapper=True) as map_file:
+            mask_map_info = map_file.read_map_info("mask")
+        self.nside = self.lmax = mask_map_info["nside"]
+        self.nest = mask_map_info["nest"]
+
+        self.set_z_shells()
+
+        self.generate_shell_cls()
+
+        self.generate_binned_cls()
+
+        self.generate_catalogs()
+
+    def set_z_shells(self):
+        """
+        Set up redshift shells with uniform spacing in comoving distance
+
+        These shells determine the redshift resolution of the simulation so should
+        be narrower than the n(z)s
+
+        This method initializes redshift bins based on the provided configuration
+        options. It calculates redshift shells and associated weight functions
+        for later use
+
+        Uses CCl rather than glass.shells.distance_grid to avoid mixing CCL and CAMB
+        """
+        import glass.shells
+
+        with self.open_input("fiducial_cosmology", wrapper=True) as f:
+            cosmo = f.to_ccl()
+
+        chi_min = cosmo.comoving_radial_distance(1 / (1 + self.config["zmin"]))
+        chi_max = cosmo.comoving_radial_distance(1 / (1 + self.config["zmax"]))
+        chi_grid = np.arange(chi_min, chi_max + self.config["dx"], self.config["dx"])
+        a_grid = cosmo.scale_factor_of_chi(chi_grid)
+        self.zb = (1.0 / a_grid) - 1.0
+        self.ws = glass.shells.tophat_windows(self.zb, weight=camb_tophat_weight)
+        self.nshells = len(self.ws)
+        # TO DO: figure out why GLASS needed the linear ramp weight here
+
+    def get_bias_at_z(self, z):
+        bias0 = self.config["bias0"]
+        alpha_bz = self.config["alpha_bz"]
```

Comment:
Where is `bias0`, `alpha_bz`, and `zpivot` defined for the configs?

Reply:
I added an example in examples/lognormal to show how the configs should be designed

Reply:
okay great! my main worry is that there aren't default values given, so what happens if none are provided? will the stage crash?

Reply:
yes it will return an error. I was reluctant to put in defaults because it will be very sample dependent that so someone using the stage probably *should* have to think about. happy to just put b0=2 as default though if you think thats better

Reply:
I think it might be better, and then add a big comment that people really should be adjusting these! that way people can at least go to the config list and see all the configs defined. But if you and Joe prefer it otherwise then leave it as is

Reply:
sounds good. I added the defaults